### ISSUE TITLE
Support CEF 3.3202.1686.gd665578 / Chromium 62

### ISSUE THREAD

**Issue by perlun** [cefsharp/CefSharp#2200]
This is an umbrella release to support the Chromium 62 version in CefSharp, which I think a lot of you are waiting for.

Some of the known issues thus far:

- [x] ~https://github.com/cefsharp/cef-binary/pull/46 - publish cef.redist and cef.sdk packages. This work has been started by @peters but is not fully completed yet. (and his PR is using an old CEF version as well.)~ I think we take the simpler approach in CefSharp/cef-binary#53 for now.
- [x] Upgrade `cef.sdk` dependency in CefSharp, making required changes as needed - merged in #2203.
- [ ] #1203 - JS bindings are no longer working as expected. **Fixing this might be postponed to post-62.0.0 release**, see [this comment](https://github.com/cefsharp/CefSharp/issues/2200#issuecomment-352206491) and use the 👍 and 👎 to cast your vote.
- [ ] Perhaps other things, not yet noted.

Fixes we get for free because of the upgrade:

- Closes #1916: PDF scrolling with mouse wheel not working
- Closes #2170: BrowserSubprocess memory leak + crash.
- Closes #2182: WebGL software rendering is broken in CefSharp 57
- Probably a bunch of others, see https://github.com/cefsharp/CefSharp/releases/tag/v62.0.0-pre01 for the full release notes.

----

**Update**: Prerelease packages are available on nuget.org, please ensure to check the "Include prereleases" checkbox in the NuGet settings in Visual Studio.

**Comment by mjbrookecub** [comment]
Hiya, any ideas on when this is likely to be ready? I've found a bug with the version of cef that v57 is using and am looking forward to that being resolved.

**Comment by perlun** [comment]
> Hiya, any ideas on when this is likely to be ready?

ASAP. 😄 Hopefully rather soon, but because of the nature of the project it's difficult to make promises at this stage. We'll let you and everyone else eagerly waiting know as soon as it's ready, for sure.

**Comment by SilverDragon135** [comment]
CEF 62 probably fixes #1916 too.
I had scrolling problems in winforms too in case I was scrolling during the loading - fixed with CEF 62 for me.

**Comment by perlun** [comment]
Thanks @SilverDragon135, added it to the list.

**Comment by perlun** [comment]
[Chrome 63](https://chromereleases.googleblog.com/2017/12/stable-channel-update-for-desktop.html) was recently released. We should probably re-target this to go with CEF 63 instead.

If someone wants to do the initial work of bringing it into cef-binary, by all means, be my guest!

**Comment by muffmolch** [comment]
thus.. no cefsharp for 62?

**Comment by perlun** [comment]
> thus.. no cefsharp for 62?

Not yet, no. We could make a release but it still has the #1203 bug present which will prevent some people from using it.

Please vote on this comment: 👎 if you want to have a 62 release _asap_, or 👍 if you'd rather wait some extra time but get a 63 release instead. Nothing is free in this world, everything takes a bit of time and it would be nice to hear your feedback before spending time on this.

I will base my decision on your feedback. I personally would prefer the 63 approach (since it seems silly to release a version which is already outdated at the time of release), but I'm very open to your opinions also.

**Comment by Cansii** [comment]
Current Cefsharp release version (57) have stability issue (probably #2170) so we got many stabily issues on our latest product version. It may leads to security issue for ours systems, so I realy hope you can release new stable version soon.
If not, I have to revert to last realy stable release (53 for me).
Thanks a lot for your great job.

**Comment by muffmolch** [comment]
question is: is version 63 really uptodate when the new CEFVersion gets released ;-)

**Comment by perlun** [comment]
@Cansii 

> Current Cefsharp release version (57) have stability issue (probably #2170) so we got many stabily issues on our latest product version. It may leads to security issue for ours systems, so I realy hope you can release new stable version soon.

Stability issues is one thing, I know about #2170, but _security_ issues? Are we not mixing things up a bit here?

I understand if people have problems with 57, that's fine. I think bumping the release to use 63 doesn't really have to be that much work to be honest though, so I am leaning towards that route since everyone will thank me for doing so in the long run... 😉

**Comment by Cansii** [comment]
@perlun 

> Stability issues is one thing, I know about #2170, but security issues? Are we not mixing things up a bit here?

As I said, stabilty issue leads to some security issues in my product. Uncatchable exceptions lead to a app crash. And my app need to be running continuously (to ensure kiosk mode).

So, I probably have to reingeneer my app architecture, but this is not a mixing...

I agree, in the long run, 63 is better. 

You know the time cost of this choice. But you ask, so I give you my anser. ;)

**Comment by perlun** [comment]
> You know the time cost of this choice. But you ask, so I give you my anser. ;)

That's fine. 😄 I understand your concerns, thanks for using CefSharp.

I noticed that CEF 63 isn't yet available at http://opensource.spotify.com/cefbuilds/index.html, so let's go with a 62 prerelease instead. Will try to get it done within the next few days.

**Comment by Cansii** [comment]
> Will try to get it done within the next few days.
Great !

Maybe could you give me a trick about a way to catch this uncatchable exception ?

**Comment by perlun** [comment]
@Cansii Let's move that discussion to Gitter or a separate issue. If it's exceptions being raised on the C++ side, I think the unfortunate answer is that they cannot be caught. But, as stated, let's move that part of the discussion elsewhere.

**Comment by perlun** [comment]
@Cansii and others - a first 62.0.0 prerelease is now available: https://github.com/cefsharp/CefSharp/releases/tag/v62.0.0-pre01

Please test it and give us your feedback here in this issue.

**Comment by mjbrookecub** [comment]
It seems to work really well for me. Solves two of the issues I had with v57

**Comment by muffmolch** [comment]
It's fine on our side, too! Thanks a lot!

**Comment by vwmberry95** [comment]
Any plans for another build with proprietary codecs?  I have "62.0.0-CI2368" installed, if I try to update CefSharp.Common and CefSharp.Wpf to "62.0.0-pre01" I get this message:

Unable to resolve dependencies. 'cef.redist.x86 3.3202.1690-proprietary-codecs' is not compatible with 'CefSharp.Common 62.0.0-pre01 constraint: cef.redist.x86 (= 3.3202.1686)'.

I can roll with "62.0.0-CI2368" as it's working great, I was just hoping to test "62.0.0-pre01" for you as well while keeping H.264 support.

Greatly appreciated as always!

**Comment by mitchcapper** [comment]
@vwmberry95 there are potential licensing issues with proprietary codecs so it will likely not be in the official tree.  You can compile from scratch yourself these features in although it is tedious.    As for the errors if you run into it on the current nuget packages it may be due to installing the former 62 test which had those dependencies and left elements in your project file.  Remove cefsharp from your project and save the project file.  Open the project file in a text editor and remove any cef references (you may find older versions as well).  Then installing 62.0.0-pre01 should work.

**Comment by vwmberry95** [comment]
@mitchcapper Thanks, mitch.  Yeah, I should have targeted that comment towards @perlun as he did the build with codecs under the myget tree due to licensing as you said.  Being so tedious, having it done once in a fashion that correlates with the CefSharp release would be greatly appreciated by many, I'm sure.

That being said, I did try building from scratch in the past but had issues, perhaps because CefSharp was still targeting v57 at the time.  Even building cef at the same version had issues.

**Comment by perlun** [comment]
> @mitchcapper Thanks, mitch. Yeah, I should have targeted that comment towards @perlun as he did the build with codecs under the myget tree due to licensing as you said. Being so tedious, having it done once in a fashion that correlates with the CefSharp release would be greatly appreciated by many, I'm sure.

Yeah, like @mitchcapper mentioned I think it will be hard to include that code in the official tree (unfortunately!). What we _could_ perhaps do is to make the official packages be runnable towards either version of the `cef.redist` or something... I don't know. Maybe something like Debian did with their "non-US" files years ago (if anyone remember, back in the days when cryptography was very patent/etc-encumbered... before the US relaxed these restrictions.)

(In other words: I don't have a clear plan for this at the moment. Right now #1203 _kind_ of blocks us from doing a stable CefSharp 62 release anyway, so you can choose between "two different prereleases" at the moment - one "official" and one "unofficial".)

**Comment by gusowski1** [comment]
@perlun why does #1203 blocks the release?
As far as i see is issue #1203 open since 2015 - wouldn't it be better to release CefSharp 62 from an security aspect and fix #1203 afterwards and make a 62. release (or 63)

**Comment by perlun** [comment]
> @perlun why does #1203 blocks the release?

Because we anticipate so many people would upgrade to 62.0.0 and have problems because of it, since people (unfortunately) tend to upgrade without reading release notes carefully enough, etc...

So the expectancy is that it will cause more problems for people than it would solve.

> As far as i see is issue #1203 open since 2015 - wouldn't it be better to release CefSharp 62 from an security aspect and fix #1203 afterwards and make a 62. release (or 63)

The long story:

- We have had a workaround in place since September 24 2015 via dcf4eb4 (fixed by @amaitland)
- In CEF 3029 (CEF 58), the workaround stopped working because of changes upstream.
- This is one of the main reasons why there hasn't been any stable release of CefSharp since CefSharp 57.0.0. We haven't been able to implement a proper workaround for this yet, unfortunately. (mostly because of lack of time/resources to work on the project.)

If all goes well and @amaitland gets enough funding, #1203 will be sorted out in the quite near future: https://gist.github.com/amaitland/91dae9ecfe8c47d0da08e000685eeef2

----

We can always have a vote: use the 👍 on this comment if you want to see us release a 62.0.0 asap _with the bug present_ (and perhaps a setting flag like Alex suggested to make people be aware of it and "opt in" to "JS bindings with defect").

Use the 👎 if you rather prefer to wait on a 62.0.0 or 63.0.0 _until the bug is fixed_. We do take our user feedback very seriously, and if we get a reasonable enough amount of feedback on this, we will base the decision on this feedback. Thanks!

**Comment by mrtylerzhou** [comment]
63 is available now

**Comment by Cansii** [comment]
> @perlun  - a first 62.0.0 prerelease is now available: 
> Please test it and give us your feedback here in this issue.

So I made a test with pre-release nuget package :

- x64
- Chromium 62.0.3202.94
- CEF : r3.3202.1686.gd665578
- CefSharp : 59.0.0.0
 
**and catching UnhandledException** with UnhandledExceptionEventHandler to restart my app.

It seems really better for last 12 hours. But I will wait some times before putting it on most critical install.

And **I've got an issue** on a particular install _(it does not happen on mine)_ on a particular page with a video. The webm video plays well but the rest of the page is'nt displayed (maybe hidden). I've enought time to get more details at this time. I will investigate this as soon as possible.
